state.go

```diff
@@ -27,76 +31,804 @@ type State struct {
+	}
+
+	updateProposal := Proposal{
+		Update: &UpdateProposal{
+			LeafKey: key.PublicKey,
+		},
+	}
+
+	return s.sign(updateProposal)
+}
+
+func (s *State) remove(removed leafIndex) *MLSPlaintext {
+	removeProposal := Proposal{
+		Remove: &RemoveProposal{
+			Removed: uint32(removed),
+		},
+	}
+	return s.sign(removeProposal)
+}
+
+func (s *State) commit(leafSecret []byte) (*MLSPlaintext, *Welcome, *State, error) {
+	commit := Commit{}
+	var joiners []ClientInitKey
+
+	for _, pp := range s.PendingProposals {
+		pid := s.proposalId(pp)
+		if s.processedProposals[pid.String()] {
+			fmt.Printf("mls.state: skipping processed proposal %v", pp)
+			continue
+		}
+
+		proposal := pp.Content.Proposal
+		switch {
+		case proposal.Add != nil:
+			commit.Adds = append(commit.Adds, pid)
+			joiners = append(joiners, proposal.Add.ClientInitKey)
+		case proposal.Update != nil:
+			commit.Updates = append(commit.Updates, pid)
+		case proposal.Remove != nil:
+			commit.Removes = append(commit.Removes, pid)
+		}
+	}
+
+	// init new state to apply commit and ratchet forward
+	next := s.clone()
+	next.apply(commit)
+	next.PendingProposals = nil
+
+	// Start a GroupInfo with the prepared state
+	prevInitSecret := s.Keys.InitSecret
+	gi := new(GroupInfo)
+	gi.GroupId = next.GroupID
+	gi.Epoch = next.Epoch + 1
+	gi.Tree = next.Tree.clone()
+	gi.PriorConfirmedTranscriptHash = s.ConfirmedTranscriptHash
+
+	ctx, err := syntax.Marshal(GroupContext{
+		GroupID:                 gi.GroupId,
+		Epoch:                   gi.Epoch,
+		TreeHash:                gi.Tree.RootHash(),
+		ConfirmedTranscriptHash: gi.PriorConfirmedTranscriptHash,
+	})
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("mls.state: grpCtx marshal failure %v", err)
+	}
+
+	// KEM new entropy to the group and the new joiners
+	path, updateSecret := next.Tree.Encap(s.Index, ctx, leafSecret)
+
+	// Create the Commit message and advance the transcripts / key schedule
+	pt := next.ratchetAndSign(commit, updateSecret, s.groupContext())
+
+	// Complete the GroupInfo and form the Welcome
+	gi.ConfirmedTranscriptHash = next.ConfirmedTranscriptHash
+	gi.InterimTranscriptHash = next.InterimTranscriptHash
+	gi.Path = path
+	gi.Confirmation = pt.Content.Commit.Confirmation
+	err = gi.sign(s.Index, &s.IdentityPriv)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("mls.state: groupInfo sign failure %v", err)
+	}
+
+	welcome := newWelcome(s.CipherSuite, prevInitSecret, gi)
+	for _, joiner := range joiners {
+		welcome.encrypt(joiner)
+	}
+
+	return pt, welcome, next, nil
+}
+
+/// Proposal processing helpers
+
+func (s *State) apply(commit Commit) {
+	s.applyProposals(commit.Updates)
+	s.applyProposals(commit.Removes)
+	s.applyProposals(commit.Adds)
+	// tree truncate
+
+}
+
+func (s *State) applyAddProposal(add *AddProposal) {
+	target := s.Tree.LeftmostFree()
+	s.Tree.AddLeaf(target, &add.ClientInitKey.InitKey, &add.ClientInitKey.Credential)
```

Comment:
Seems like this needs error propagation, since there's clearly at least one error case (when the target is occupied)